File: solutions/javascript/knapsack/1/knapsack.js
//
// This is only a SKELETON file for the 'Knapsack' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const knapsack = (maximumWeight, items) => {
 if (maximumWeight === 0 || items.length === 0) return 0;
   const validItems = items.filter(item => item.weight <= maximumWeight);
   if (validItems.length === 0) return 0;

  const dp = Array(maximumWeight + 1).fill(0);

  for (let item of validItems) {
    for (let w = maximumWeight; w >= item.weight; w--) {
      dp[w] = Math.max(dp[w], dp[w - item.weight] + item.value);
    }
  }

  return dp[maximumWeight];

};

File: solutions/python/knapsack/1/knapsack.py

from itertools import combinations

def maximum_value(maximum_weight, items):
    

    best_value = 0

    n = len(items)

    for i in range(n + 1):

        combos = combinations(items, i)

        for combo in combos:

            total_weight = sum(item["weight"] for item in combo)
            total_value  = sum(item["value"]  for item in combo)

            if total_weight <= maximum_weight:
                best_value = max(best_value, total_value)

    return best_value
